src/coreclr/jit/decomposelongs.cpp

```diff
@@ -1835,6 +1873,93 @@ GenTree* DecomposeLongs::DecomposeHWIntrinsicGetElement(LIR::Use& use, GenTreeHW
 }
 
+//------------------------------------------------------------------------
+// DecomposeHWIntrinsicToScalar: Decompose GT_HWINTRINSIC -- NI_Vector*_ToScalar.
+//
+// create:
+//
+// tmp_simd_var = simd_var
+// lo_result = GT_HWINTRINSIC{ToScalar}[int](tmp_simd_var)
+// hi_result = GT_HWINTRINSIC{GetElement}[int](tmp_simd_var, 1)
+//             - or -
+//             GT_HWINTRINSIC{ToScalar}[int](GT_RSZ(tmp_simd_var, 32))
+// return: GT_LONG(lo_result, hi_result)
+//
+// Arguments:
+//    use - the LIR::Use object for the def that needs to be decomposed.
+//   node - the hwintrinsic node to decompose
+//
+// Return Value:
+//    The GT_LONG node wrapping the upper and lower halves.
+//
+GenTree* DecomposeLongs::DecomposeHWIntrinsicToScalar(LIR::Use& use, GenTreeHWIntrinsic* node)
+{
+    assert(node == use.Def());
+    assert(varTypeIsLong(node));
+    assert(HWIntrinsicInfo::IsVectorToScalar(node->GetHWIntrinsicId()));
+
+    GenTree*       op1          = node->Op(1);
+    NamedIntrinsic intrinsicId  = node->GetHWIntrinsicId();
+    var_types      simdBaseType = node->GetSimdBaseType();
+    unsigned       simdSize     = node->GetSimdSize();
+
+    assert(varTypeIsLong(simdBaseType));
+    assert(varTypeIsSIMD(op1));
+
+    GenTree* simdTmpVar    = RepresentOpAsLocalVar(op1, node, &node->Op(1));
+    unsigned simdTmpVarNum = simdTmpVar->AsLclVarCommon()->GetLclNum();
+    JITDUMP("[DecomposeHWIntrinsicToScalar]: Saving op1 tree to a temp var:\n");
+    DISPTREERANGE(Range(), simdTmpVar);
+
+    GenTreeHWIntrinsic* loResult =
+        m_compiler->gtNewSimdHWIntrinsicNode(TYP_INT, simdTmpVar, intrinsicId, CORINFO_TYPE_INT, simdSize);
+    Range().InsertAfter(simdTmpVar, loResult);
+
+    simdTmpVar = m_compiler->gtNewLclLNode(simdTmpVarNum, simdTmpVar->TypeGet());
+    Range().InsertAfter(loResult, simdTmpVar);
+
+    GenTreeHWIntrinsic* hiResult;
+    if (m_compiler->compOpportunisticallyDependsOn(InstructionSet_SSE41))
+    {
+        NamedIntrinsic getElement = NI_Illegal;
+        switch (simdSize)
+        {
+            case 16:
+                getElement = NI_Vector128_GetElement;
+                break;
+            case 32:
+                getElement = NI_Vector256_GetElement;
+                break;
+            case 64:
+                getElement = NI_Vector512_GetElement;
+                break;
+            default:
+                unreached();
+        }
+
+        GenTree* one = m_compiler->gtNewIconNode(1);
+        hiResult =
+            m_compiler->gtNewSimdHWIntrinsicNode(TYP_INT, simdTmpVar, one, getElement, CORINFO_TYPE_INT, simdSize);
+
+        Range().InsertAfter(simdTmpVar, one, hiResult);
+    }
+    else
+    {
+        assert(m_compiler->compIsaSupportedDebugOnly(InstructionSet_SSE2));
+
+        GenTree* thirtyTwo = m_compiler->gtNewIconNode(32);
+        GenTree* shift     = m_compiler->gtNewSimdBinOpNode(GT_RSZ, op1->TypeGet(), simdTmpVar, thirtyTwo,
```

Comment:
Isn't this missing `ToScalar()` to get the 32-bit integer out of the simd result?

Reply:
ToScalar is the original `intrinsicId`, and it's built on the next line down.  Full SSE2 codegen for `Vector128<ulong>.ToScalar()`:

```asm
; Method Program:ToScalar(System.Runtime.Intrinsics.Vector128`1[ulong]):ulong (FullOpts)
G_M34649_IG01:  ;; offset=0x0000
       push     ebp
       mov      ebp, esp
       movups   xmm0, xmmword ptr [ebp+0x08]
						;; size=7 bbWeight=1 PerfScore 4.25

G_M34649_IG02:  ;; offset=0x0007
       movd     eax, xmm0
       psrlq    xmm0, 32
       movd     edx, xmm0
						;; size=13 bbWeight=1 PerfScore 4.50

G_M34649_IG03:  ;; offset=0x0014
       pop      ebp
       ret      16
						;; size=4 bbWeight=1 PerfScore 2.50
; Total bytes of code: 24
```

Reply:
If we know its definitively `ToScalar` we should probably assert that and use `gtNewSimdToScalar` node instead of manually constructing such a node

Reply:
Done. It was already asserted (line 1899), but the helpers do look much better.